pkg/controller/seed-controller-manager/ipam/controller.go

```diff
@@ -0,0 +1,272 @@
+		ipamPool := a.(*kubermaticv1.IPAMPool)
+
+		clusterList := &kubermaticv1.ClusterList{}
+		if err := mgr.GetClient().List(context.Background(), clusterList); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to list Clusters: %w", err))
+			log.Errorw("Failed to list clusters", zap.Error(err))
+			return []reconcile.Request{}
+		}
+
+		requests := []reconcile.Request{}
+		for _, cluster := range clusterList.Items {
+			_, isClusterDCConfiguredInPool := ipamPool.Spec.Datacenters[cluster.Spec.Cloud.DatacenterName]
+			if !isClusterDCConfiguredInPool {
+				// The IPAM pool is not relevant to this cluster, so skip it
+				continue
+			}
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: cluster.Name}})
+		}
+		return requests
+	})
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.IPAMPool{}}, enqueueClustersForIPAMPool); err != nil {
+		return fmt.Errorf("failed to create watch for IPAM Pools: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("Skipping because the cluster is already gone")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionIPAMControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(ctx, cluster)
+		},
+	)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// List IPAM Pools
+	ipamPoolList := &kubermaticv1.IPAMPoolList{}
+	err := r.Client.List(ctx, ipamPoolList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list IPAM pools: %w", err)
+	}
+
+	// Loop IPAM pools, considering only the relevant ones (i.e. IPAM pools with same cluster datacenter)
+	for _, ipamPool := range ipamPoolList.Items {
+		clusterDC := cluster.Spec.Cloud.DatacenterName
+
+		dcIPAMPoolCfg, isClusterDCConfigured := ipamPool.Spec.Datacenters[clusterDC]
+		if !isClusterDCConfigured {
+			// This IPAM pool is not relevant to cluster, so skip it
+			continue
+		}
+
+		ipamAllocation := &kubermaticv1.IPAMAllocation{}
+		err = r.Client.Get(ctx, types.NamespacedName{Namespace: fmt.Sprintf("cluster-%s", cluster.Name), Name: ipamPool.Name}, ipamAllocation)
```

Comment:
Please take the cluster namespace from `cluster.Status.NamespaceName`

Reply:
You then also need to add an if statement to your Reconcile() to skip reconciling if the namespacename is not yet set.